### ISSUE TITLE
[bug] RL trainers bypass DDP: gradients are never all-reduced in GRPO / PPO / Agent training

### ISSUE THREAD

**Issue by basil-k-aji-dev** [jingyaogong/minimind#853]
## Summary

`trainer/train_grpo.py`, `trainer/train_ppo.py` and `trainer/train_agent.py` wrap the policy in `DistributedDataParallel`, but then run the **loss-bearing forward pass through `.module`**, bypassing the wrapper. As a result `backward()` performs no gradient all-reduce, and multi-GPU RL training silently degrades to single-GPU training on 1/N of the data.

The supervised trainers are not affected.

## Where

```python
# trainer/train_grpo.py:98
model_unwrapped = model.module if isinstance(model, DistributedDataParallel) else model
with autocast_ctx:
    res = model_unwrapped(outputs, attention_mask=full_mask)
...
loss.backward()
```

Same pattern at `trainer/train_ppo.py:162-175` (`actor_unwrapped`, `critic_unwrapped`) and `trainer/train_agent.py:281-283`.

## Cause

DDP's gradient all-reduce is driven by autograd hooks that are armed inside `DistributedDataParallel.forward()` (via `Reducer::prepare_for_backward`). When the forward goes straight to `.module`, those hooks are never armed for that iteration, so the subsequent `backward()` all-reduces nothing. Each rank then steps its optimizer on purely local gradients and the replicas diverge. Only rank 0 is ever saved, so the failure is completely silent — no error, no warning, just a model trained on a fraction of the data.

## Reproduction

Two gloo ranks on CPU, identical init, deliberately different data per rank:

```python
import torch, torch.nn as nn, torch.distributed as dist
from torch.nn.parallel import DistributedDataParallel as DDP

dist.init_process_group("gloo")
rank, ws = dist.get_rank(), dist.get_world_size()
torch.manual_seed(0)
ddp = DDP(nn.Linear(4, 1, bias=False))
x = torch.full((2, 4), float(rank + 1))
y = torch.zeros(2, 1)

for mode in ("ddp.module(x)", "ddp(x)"):
    ddp.zero_grad(set_to_none=True)
    callable_ = ddp.module if mode == "ddp.module(x)" else ddp
    ((callable_(x) - y) ** 2).mean().backward()
    g = ddp.module.weight.grad.detach().clone()
    buf = [torch.zeros_like(g) for _ in range(ws)]
    dist.all_gather(buf, g)
    if rank == 0:
        print(mode, "ranks agree =", torch.allclose(buf[0], buf[1]),
              buf[0].flatten().tolist(), buf[1].flatten().tolist())
dist.destroy_process_group()
```

```
ddp.module(x)   ranks agree = False   [-1.03]      [-4.1201]
ddp(x)          ranks agree = True    [-2.5751]    [-2.5751]
```

`-2.5751` is exactly the mean of `-1.03` and `-4.1201` — the all-reduce that the current call path never performs.

## Why this looks unintended

`train_ppo.py` already contains two safeguards that only make sense if backward drives DDP collectives:

```python
# 同步各卡的 approx_kl，防止某卡 break 而其它卡继续导致 DDP 死锁
# 早停时必须保证 forward-backward 闭环，故只截断 loss 不中断 DDP 通信
```

The code carefully keeps the collectives balanced across ranks while the mechanism that would produce them is bypassed.

Additionally, `train_pretrain.py`, `train_full_sft.py`, `train_lora.py` and `train_distillation.py` all call the wrapped `model(...)`. Only the three RL trainers deviate.

## Expected

The forward pass whose graph is backpropagated should go through the DDP-wrapped module. `no_grad` uses of `.module` (rollout generation, checkpoint saving) are correct as-is.

Single-GPU runs are unaffected either way: with no process group initialised, `model` is the bare module.

## Environment

- current `main` (`a3c7b01`)
- reproduced with torch 2.14.0+cpu, gloo backend, 2 ranks
